cmd/forwarder/fixer.go

```diff
@@ -63,3 +67,68 @@ func fix(r io.Reader, remoteAddr string, logplexDrainToken string) ([]byte, erro
 
 	return messageLenWriter.Bytes(), lp.Err()
 }
+
+// msgpackToSyslog converts post data msgpack messages to length prefixed syslog frames
+func msgpackToSyslog(r io.Reader, remoteAddr string, logplexDrainToken string) ([]byte, error) {
+	var messageWriter bytes.Buffer
+	var messageLenWriter bytes.Buffer
+
+	// LEN SP <PRI>VERSION SP TIMESTAMP SP HOSTNAME SP APP-NAME SP PROCID SP MSGID SP STRUCTURED-DATA MSG
+	dec := decoder.NewDecoder(r)
+	for {
+		rec, err := dec.GetRecord()
+		if err == io.EOF {
+			break // No more records
+		}
+
+		d, err := decoder.ExtractData(rec)
+		if err != nil {
+			return nil, err
+		}
+
+		ts, err := decoder.ExtractTime(rec)
+		if err != nil {
+			return nil, err
+		}
+
+		messageWriter.WriteString("<")
+		messageWriter.WriteString(fetchValues(d, "PRIORITY"))
+		messageWriter.WriteString(">1")
+		messageWriter.WriteString(" ")
+		messageWriter.WriteString(ts.UTC().Format(logplexBatchTimeFormat))
+		messageWriter.WriteString(" ")
+		if logplexDrainToken != "" {
+			messageWriter.WriteString(logplexDrainToken)
+		} else {
+			messageWriter.WriteString(fetchValues(d, "_HOSTNAME", "HOSTNAME"))
+		}
+		messageWriter.WriteString(" ")
+		messageWriter.WriteString(fetchValues(d, "SYSLOG_IDENTIFIER", "_COMM"))
+		messageWriter.WriteString(" ")
+		messageWriter.WriteString(fetchValues(d, "_PID"))
+		messageWriter.WriteString(" ")
+		messageWriter.WriteString(fetchValues(d, "MESSAGE_ID"))
+		messageWriter.WriteString(" ")
+		messageWriter.WriteString("[origin ip=\"")
+		messageWriter.WriteString(remoteAddr)
+		messageWriter.WriteString("\"]")
+		messageWriter.WriteString(" ")
+		messageWriter.WriteString(fetchValues(d, "MESSAGE"))
+		messageWriter.WriteString("\n")
+		messageLenWriter.WriteString(strconv.Itoa(messageWriter.Len()))
+		messageLenWriter.WriteString(" ")
+		messageWriter.WriteTo(&messageLenWriter)
+	}
+
+	return messageLenWriter.Bytes(), nil
+}
+
+func fetchValues(data map[string]interface{}, fields ...string) string {
+	for _, f := range fields {
+		if _, ok := data[f]; ok {
+			return data[f].(string)
```

Comment:
is there any concern about this string casting failing?

Reply:
if the value can't be cast as a string, you'll get a panic

Reply:
No concern because the keys we're fetching here will always contain strings. However there can be keys (such as the KUBERNETES key) in the payload that we're not fetching yet that hold a pointer to another `map[string]interface{}`.